src/transforms/javascript.rs

```diff
@@ -0,0 +1,789 @@
+#[serde(deny_unknown_fields)]
+pub struct JavaScriptConfig {
+    source: Option<String>,
+    path: Option<String>,
+    handler: Option<String>,
+    memory_limit: Option<usize>,
+}
+
+#[typetag::serde(name = "javascript")]
+impl TransformConfig for JavaScriptConfig {
+    fn build(&self) -> Result<Box<dyn Transform>, String> {
+        JavaScript::new(
+            self.source.clone(),
+            self.path.clone(),
+            self.handler.clone(),
+            self.memory_limit.clone(),
+        )
+        .map(|js| -> Box<dyn Transform> { Box::new(js) })
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+enum ProcessorInput {
+    Event(Event),
+    Stop,
+}
+
+enum ProcessorOutput {
+    Start(Result<(), String>),
+    Events(Vec<Event>),
+}
+
+pub struct JavaScript {
+    input: mpsc::SyncSender<ProcessorInput>,
+    output: mpsc::Receiver<ProcessorOutput>,
+}
+
+impl JavaScript {
+    pub fn new(
+        source: Option<String>,
+        path: Option<String>,
+        handler: Option<String>,
+        memory_limit: Option<usize>,
+    ) -> Result<Self, String> {
+        let (input, thread_input) = mpsc::sync_channel(0);
+        let (thread_output, output) = mpsc::sync_channel(0);
+        thread::spawn(move || {
+            let processor = JavaScriptProcessor::new(source, path, handler, memory_limit);
+            let processor = match processor {
+                Ok(processor) => {
+                    thread_output.send(ProcessorOutput::Start(Ok(()))).unwrap();
+                    processor
+                }
+                Err(e) => {
+                    return thread_output.send(ProcessorOutput::Start(Err(e))).unwrap();
+                }
+            };
+            for event in thread_input {
+                match event {
+                    ProcessorInput::Event(event) => {
+                        let mut output_events = Vec::new();
+                        let result = processor.process(&mut output_events, event);
+                        if let Err(err) = result {
+                            eprintln!("Error in JavaScript transform; discarding event\n{}", err)
+                        }
+                        thread_output
+                            .send(ProcessorOutput::Events(output_events))
+                            .unwrap();
+                    }
+                    ProcessorInput::Stop => break,
+                }
+            }
+        });
+        match output.recv().unwrap() {
+            ProcessorOutput::Start(res) => res.map(|_| JavaScript { input, output }),
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl Drop for JavaScript {
+    fn drop(&mut self) {
+        self.input.send(ProcessorInput::Stop).unwrap();
+    }
+}
+
+impl Transform for JavaScript {
+    // only used in tests
+    fn transform(&mut self, event: Event) -> Option<Event> {
```

Comment:
In other transforms which don't multiply events only `transform` function is implemented instead of `transform_into` and the [default implementation of `transform_into` from `Transform`](https://github.com/timberio/vector/blob/master/src/transforms/mod.rs#L18) trait uses it. So it cannot be put under `#[cfg(test)]` just here without reworking all other transforms.

The comment is more about this implementation being acceptable because it is used only in tests, so that it is fine to have it implemented using intermediate `Vec` and crash when there is more than one event in the output.